### PR TITLE
Add a timeout option for verification of synthesized terms.

### PR DESCRIPTION
**Pull request by abdoo8080**
This PR adds an option to timeout on the verification check for the terms enumerated by the Sygus solver.

### CODE REVIEW

File: src/options/quantifiers_options.toml [hunk 1]
```diff
@@ -1318,6 +1318,13 @@ name   = "Quantifiers"
   default    = "3"
   help       = "maximum number of instantiation rounds for sygus verification calls (-1 == no limit, default is 3)"
 
+[[option]]
+  name       = "sygusVerifyTimeout"
+  category   = "regular"
+  long       = "sygus-verify-timeout=N"
+  type       = "int64_t"
```
Comment (nafur) [inline src/options/quantifiers_options.toml:1325]:
```suggestion
  type       = "uint64_t"
  default   = 0
```
the time limit is a `uint64_t` internally. Also, to make this explicit and to easily show it in the documentation, specifying defaults have recently become mandatory.

File: src/theory/quantifiers/sygus/synth_verify.cpp [hunk 2]
```diff
@@ -110,7 +110,17 @@ Result SynthVerify::verify(Node query,
   }
   Trace("sygus-engine") << "  *** Verify with subcall..." << std::endl;
   query = rewrite(query);
-  Result r = checkWithSubsolver(query, vars, mvs, d_subOptions, d_subLogicInfo);
+  Result r;
+  if (options().quantifiers.sygusVerifyTimeoutWasSetByUser)
```
Comment (nafur) [inline src/theory/quantifiers/sygus/synth_verify.cpp:114]:
```suggestion
  if (options().quantifiers.sygusVerifyTimeout != 0)
```
This captures the actual intention, I guess: The user may explicitly set it to zero, or the default may become something else at some point.

Comment (abdoo8080) [reply src/theory/quantifiers/sygus/synth_verify.cpp]: Good point!

**Review by nafur — COMMENTED** [review]
I've just had a look at the overall timeout logic for `checkWithSubsolver`.
For once, I think it can be simplified significantly:
- remove the `needsTimeout` flag and simply check `timeout > 0` instead.
- in this PR, remove the case distinction and simply use the first variant

This leaves one issue, where I'm not sure whether it is a problem: right now, if `sygusVerifyTimeout` is zero, the general timeout is copied. This sounds reasonable to me.
If it is set, though, it "extends" the timeout beyond the timeout of the main solver. Is this what we want? This is obvious when `sygusVerifyTimeout > perCallMillisecondLimit`. But the mere fact of calling a subsolver means that runs with a new `ResourceManager` which begins counting resources from scratch...
Note, though, that this is a more general thing to think about... @ajreynol

**Comment by ajreynol** [comment]
I don't think its all that much simpler eliminating the `needsTimeout` flag, although I do agree it would be best if the 2 branches in this PR were merged.  We can already do this right?  By passing the condition and value.

I actually do think that `options().quantifiers.sygusVerifyTimeoutWasSetByUser` is the better condition, as @abdoo8080 originally had it.  If the user sets the time limit to 0, this should mean that the time limit is 0, not unlimited.

**Comment by nafur** [comment]
> I actually do think that `options().quantifiers.sygusVerifyTimeoutWasSetByUser` is the better condition, as @abdoo8080 originally had it. If the user sets the time limit to 0, this should mean that the time limit is 0, not unlimited. This is a somewhat strange setting, but it would mean that the user wanted to enumerate terms but never verify.

Hm, I see. I'm not sure it is a good idea to "hide" the "do I want to verify them" within the timeout option. Also, I don't think it works that way right now: it simply sets the per-call time limit, which is not in effect if it is zero.

**Comment by ajreynol** [comment]
Ah, so the policy has to be same I guess is what you're saying?  I don't have a very strong opinion on this.

**Comment by nafur** [comment]
Yes, the policy for all other limits (time or resource) is that 0 means there is no limit.
You could, I guess, make this timeout be signed and add have `-1 = no checking, 0 = checking but no timeout, n>0 = checking with timeout`.

While I like trying to get away with less options, my feeling is that having `--sygus-verify` to enable/disable verification and `--sygus-verify-timeout` as separate options makes it more clear to the user.

**Comment by ajreynol** [comment]
Let's just go with `options().quantifiers.sygusVerifyTimeout != 0` as a condition then, but the branches can be merged regardless.

I can do a pass that corrects this logic for other subsolver calls, although this is minor issue.

File: src/options/quantifiers_options.toml [hunk 3]
```diff
@@ -1318,6 +1318,14 @@ name   = "Quantifiers"
   default    = "3"
   help       = "maximum number of instantiation rounds for sygus verification calls (-1 == no limit, default is 3)"
 
+[[option]]
+  name       = "sygusVerifyTimeout"
+  category   = "regular"
+  long       = "sygus-verify-timeout=N"
+  type       = "uint64_t"
+  default    = "0"
+  help       = "timeout (in milliseconds) for verifying satisfiability of synthesized terms"
```
Comment (nafur) [inline src/options/quantifiers_options.toml:1327]:
```suggestion
  help       = "timeout (in milliseconds) for verifying satisfiability of synthesized terms. setting the timeout to zero disables the verification of synthesized terms."
```

**Review by nafur — APPROVED** [review]
Just a comment about making "timeout=0 disabling the verification" explicit in the documentation of the option. Otherwise LGTM.

File: src/options/quantifiers_options.toml (hunk 3 above)
Comment (ajreynol) [reply src/options/quantifiers_options.toml:1328]: Right, I agree, it disables the timeout, the message doesnt need to change.

Comment (nafur) [reply src/options/quantifiers_options.toml:1328]: Ah, sorry. I was confused about the semantics you wanted to have in the end. You are right...

**Pull request merged**